codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerEnumGenerator.kt

```diff
@@ -60,47 +57,86 @@ open class ConstrainedEnum(
-                                context.enumTrait,
-                            ),
-                    )
-                }
+                    "EnumShapeConstraintViolationImplBlock" to
+                        validationExceptionConversionGenerator.enumShapeConstraintViolationImplBlock(
+                            context.enumTrait,
+                        ),
+                )
             }
-            rustBlock("impl #T<&str> for ${context.enumName}", RuntimeType.TryFrom) {
-                rust("type Error = #T;", constraintViolationSymbol)
-                rustBlockTemplate("fn try_from(s: &str) -> #{Result}<Self, <Self as #{TryFrom}<&str>>::Error>", *preludeScope) {
-                    rustBlock("match s") {
-                        context.sortedMembers.forEach { member ->
-                            rust("${member.value.dq()} => Ok(${context.enumName}::${member.derivedName()}),")
+        }
+
+        generateTryFromStrAndString(context)
+    }
+
+    override fun implFromForStr(context: EnumGeneratorContext): Writable =
+        generateConstraintViolation(context) {
+            rustTemplate(
+                """
+                impl #{TryFrom}<&str> for ${context.enumName} {
+                    type Error = #{ConstraintViolation};
+                    fn try_from(s: &str) -> #{Result}<Self, <Self as #{TryFrom}<&str>>::Error> {
+                        match s {
+                            #{MatchArms}
+                            _ => Err(#{ConstraintViolation}(s.to_owned()))
                         }
-                        rust("_ => Err(#T(s.to_owned()))", constraintViolationSymbol)
                     }
                 }
-            }
+                impl #{TryFrom}<#{String}> for ${context.enumName} {
+                    type Error = #{ConstraintViolation};
+                    fn try_from(s: #{String}) -> #{Result}<Self, <Self as #{TryFrom}<#{String}>>::Error> {
+                        s.as_str().try_into()
+                    }
+                }
+                """,
+                *preludeScope,
+                "ConstraintViolation" to constraintViolationSymbol,
+                "MatchArms" to
+                    writable {
+                        context.sortedMembers.forEach { member ->
+                            rust("${member.value.dq()} => Ok(${context.enumName}::${member.derivedName()}),")
+                        }
+                    },
+            )
+        }
+
+    override fun implFromForStrForUnnamedEnum(context: EnumGeneratorContext): Writable =
+        generateConstraintViolation(context) {
             rustTemplate(
                 """
+                impl #{TryFrom}<&str> for ${context.enumName} {
+                    type Error = #{ConstraintViolation};
+                    fn try_from(s: &str) -> #{Result}<Self, <Self as #{TryFrom}<&str>>::Error> {
+                        s.to_owned().try_into()
```

Comment:
This is first converting to a heap-allocated `String` and only then matching on the enum values. So invalid enum values would unnecessarily be heap-allocated. We should make `TryFrom<String>` and `TryFrom<&str>` instead delegate to `FromStr`, which should only heap-allocate when the enum value is valid.

Reply:
Unlike named enums, which can use `&str` for comparison and directly return an enum variant, unnamed enums need to store an owned, heap-allocated `String`. The `TryFrom<String>` implementation already receives an owned `String` as a parameter, so delegating to `FromStr` would result in an unnecessary heap allocation. Additionally, `ConstraintViolation` takes ownership of a heap-allocated `String`. Therefore, both code paths require a heap-allocated `String`. Calling `to_owned` within `TryFrom<&str>` only shifts the allocation slightly earlier without adding any additional allocation.